src/test/java/bio/terra/service/filedata/DrsServiceTest.java

```diff
@@ -651,14 +702,108 @@ public void postAccessUrlForObjectIdInvalidPassport() {
         UnauthorizedException.class,
         () ->
             drsService.postAccessUrlForObjectId(
-                googleDrsObjectId, "gcp-passport-us-central1", drsPassportRequestModel));
+                googleDrsObjectId, "gcp-passport-us-central1", drsPassportRequestModel, null));
+  }
+
+  private static Stream<Arguments> testSignGcpUrl() {
+    return Stream.of(
+        arguments(null, false), arguments("myProject", false), arguments("myProject", true));
+  }
+
+  @ParameterizedTest
+  @MethodSource
+  void testSignGcpUrl(String billingProject, boolean selfHostedDataset) {
+    SnapshotSummaryModel snapshotSummary =
+        new SnapshotSummaryModel().id(snapshotId).selfHosted(selfHostedDataset);
+    Snapshot snapshot =
+        mockSnapshot(snapshotId, billingProfile.getId(), CloudPlatform.GCP, SNAPSHOT_DATA_PROJECT);
+    snapshot.getSourceDataset().getDatasetSummary().selfHosted(selfHostedDataset);
+
+    Storage storage = mock(Storage.class);
+    Bucket bucket = mock(Bucket.class);
+    when(bucket.getLocation()).thenReturn("us-central1");
+    String sourcePath = googleFsFile.getCloudPath();
+
+    if (selfHostedDataset) {
+      String datasetProject = snapshot.getSourceDataset().getProjectResource().getGoogleProjectId();
+      snapshotSummary.selfHosted(selfHostedDataset);
+      when(storage.get(eq(BlobId.fromGsUtilUri(sourcePath).getBucket()), any())).thenReturn(bucket);
+      // This mock is needed in order to get the region from the bucket
+      when(gcsProjectFactory.getStorage(datasetProject)).thenReturn(storage);
+
+      when(samService.signUrlForBlob(TEST_USER, billingProject, sourcePath, URL_TTL))
+          .thenReturn(
+              "https://storage.googleapis.com/path/to/file.txt"
+                  + "?requestedBy=%s&userProject=%s"
+                      .formatted(
+                          URLEncoder.encode(TEST_USER.getEmail(), StandardCharsets.UTF_8),
+                          billingProject));
+    } else {
+      String snapshotProject = snapshot.getProjectResource().getGoogleProjectId();
+      // Note: we have to do some shenanigans with private methods in order for this to work since
+      // Google doesn't implement equals/hashcode on their {@link SignUrlOption} class
+      when(storage.signUrl(
+              eq(BlobInfo.newBuilder(GcsUriUtils.parseBlobUri(sourcePath)).build()),
+              eq(15L),
+              eq(TimeUnit.MINUTES),
+              any(),
+              any()))
+          .thenAnswer(
+              a -> {
+                Map<String, String> queryParams =
+                    TestUtils.extractValueFromPrivateObject(
+                        a.getArgument(3, Storage.SignUrlOption.class), "value");
```

Comment:
I'm trying to figure out why this is necessary. Is it to verify that something (?) is making a `SignUrlOption` with the values we're expecting? It seems like a more unit-y test would be separately checking the call to whatever's creating `SignUrlOption` if that is happening outside our code.

Reply:
I think this part of the test is doing two things. One is getting the query params from the `SignUrlOption` and one is  verifying that the correct values were passed to `Storage.SignUrlOption.withQueryParams()`. You can separate the these two concerns by having a method that returns `queryParams` in `DrsService` and one that wraps the result of that method in `Storage.SignUrlOption[]`, so you can independently test that the query params are being correctly computed.

Once you have a unit test for that, you can replace this code with passing in `billingProject` and `TEST_USER.getEmail()` directly to `formatted()` since there's no need to test the inner workings of `SignUrlOption`, and there's a downside in that testing it this way makes the test fragile if the google storage internals ever change.